k8s/clusters/README.md

```diff
@@ -0,0 +1,18 @@
+# Creating kubernetes cluster
+Kubernetes clusters are configured using kops, each cluster created right now will have 3 masters and 3 nodes in each availability zones
```

Comment:
As I've mentioned before in our meetings, it is my recommendation to run each cluster in a separate AZ, and not to spread a single cluster across multiple AZs. http://docs.heptio.com/content/aws-qs-multi-az.html gives some additional background and makes the same recommendation.

Reply:
After discussing with @jwhitlock yesterday he mentioned he was ok with this setup for now.

Reply:
Thanks for that heptio link @jgmize! @jwhitlock will make the final decision here, but based on the heptio recommendation, it seems that for now, perhaps we should go with a single AZ cluster with a single master (which is what we have now), and then we're poised to add HA later. @limed I don't think it would be that difficult to tear down the current cluster and spin-up a new one in a single AZ (with a single master), right?

Reply:
My go-ahead yesterday was based on my limited understanding of the issue, and the perception that losing a master node was difficult to recover from (which I think @jgmize talked about, but maybe that was losing the etcd cluster). I find the Heptio article pretty convincing in arguing against aiming for high-availability in the master node, but instead on automating master recovery.

I read the linked [Kubernes Federation](https://kubernetes.io/docs/concepts/cluster-administration/federation/) docs, which seem to describe the 3 master + 3 node configuration, but also suggest "maybe don't use Federation v1, wait for Federation v2, and don't ask us when v2 will ship".

Under "Scope of a single cluster":

> We suggest that all the VMs in a Kubernetes cluster should be in the same availability zone, because:
> * compared to having a single global Kubernetes cluster, there are fewer single-points of failure.
> * compared to a cluster that spans availability zones, it is easier to reason about the availability properties of a single-zone cluster.
> * when the Kubernetes developers are designing the system (e.g. making assumptions about latency, bandwidth, or correlated failures) they are assuming all the machines are in a single data center, or otherwise closely connected.

It seems that spreading nodes over several availability zones is going to have a negative impact on run-time performance, for questionable availability benefits. *sigh*.

@limed I'd like your opinion again after reading those articles, and I'm sorry if my approval sent you down the wrong path.

Reply:
It looks like responses are out of order, because I'm a very slow typer.

@limed I think I was wrong yesterday, and we should switch to a single-master, single-availability zone cluster for the staging work. There's some open questions about expanding from there (stage and prod in same cluster? second cluster in the same region? fallback cluster in a new region), but I hope we can get unstuck on CI tasks with the single cluster for staging.

Does that get you unstuck for next actions?

Reply:
I read the article and the article states

> Heptio’s recommended solution is to run a separate cluster in each Availability Zone

Which we are not planning on doing this time around, ultimately if we choose to run a single cluster on a single AZ I'm going to point out that in an event that the AZ has an outage the entire cluster goes down (Until we stand up a multi cluster single AZ setup)

AWS in generally recommends splitting things in up into multiple AZ which is why I myself would always prefer to have multi AZ single cluster. However when we do manage to get the time to setup a multi cluster single AZ setup I would be happy to follow the recommendation of the article.

Ultimately I can be swayed both ways, but if the choice is to go single AZ single cluster then I will point out that if the AZ has an outage you might lose the entire cluster

Reply:
> @limed I think I was wrong yesterday, and we should switch to a single-master, single-availability zone cluster for the staging work. There's some open questions about expanding from there (stage and prod in same cluster? second cluster in the same region? fallback cluster in a new region), but I hope we can get unstuck on CI tasks with the single cluster for staging.

Ok single cluster with single AZ it is, I don't think there are remaining questions about stage on prod since they will remain in the same cluster (They will be separated by namespaces). I don't believe we will be stuck for CI tasks if its a single cluster. The only complication for me will be when we start having multiple clusters in single AZ's since I have no experience in this setup. At that point I will just bug @jgmize to help me with that setup :)